### PR TITLE
Security proposal

### PR DESCRIPTION
**Pull request by jreidinger**
trello: https://trello.com/c/g94xQBDM/2159-8-musthave-selinux-and-security-refac

Change that unify all security settings. The change starts already by renaming firewall to security when adding cpu mitigation. Now with more features that fits into security proper code nad UI reorganization is done. Few screenshots:

summary new (see section Security)
new_summary

summary old (see section Security)
old_summary

new one with selinux configurable:
selinux_master2
new without selinux
selinux_master
and old one for comparison
selinux_old

### CODE REVIEW

File: src/lib/installation/security_settings.rb [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+        # TRANSLATORS: easy in sense the least restrictive policy
+        "easy"        => _("Easy")
+      }
+    end
+
+    # Returns a SELinux configuration handler
+    #
+    # @return [Y2Security::Selinux] the SELinux config handler
+    def selinux_config
+      require "y2security/selinux"
```
Comment (dgdavid) [inline src/lib/installation/security_settings.rb:161]: Why not put this require at the very beginning, immediately after `require 'yast'` one?

Comment (jreidinger) [reply src/lib/installation/security_settings.rb:161]: I do lazy loading for now to not care about dependencies for tests. I will need to take care about it now.

**Comment by lkocman** [comment]
Confirming that TW and Leap want to have permissive mode by default. Enforcing with imperfect policies could lead to developing habit of turning selinux off and we want to avoid that.

File: src/lib/installation/clients/security_finish.rb [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+          # The firewall will be configured according to product proposals then.
+          if Yast::Mode.auto && Y2Firewall::Clients::Auto.profile
+            log.info("Firewall: running configuration according to the AY profile")
+
+            return false unless Y2Firewall::Clients::Auto.new.write
+          else
+            Service.Enable("sshd") if @settings.enable_sshd
+            configure_firewall if @firewalld.installed?
+          end
+        end
```
Comment (dgdavid) [inline src/lib/installation/clients/security_finish.rb:73]: NP: I'd move this to its own method, `write_firewall_and_ssh_settings` or similar with its own guard clause `return if Yast::Mode.Update`

File: src/lib/installation/clients/security_finish.rb [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+            # check whether it exists
+            # give some feedback
+            # It's dozens of lines...
+            "test -x /sbin/set_polkit_default_privs && " \
+              "echo /sbin/set_polkit_default_privs && " \
+              "/sbin/set_polkit_default_privs | wc -l && " \
+              "echo 'Done'"
+          )
+          log.info "Command returned: #{ret2}"
+        end
```
Comment (dgdavid) [inline src/lib/installation/clients/security_finish.rb:103]: NP: and I'd move this to `write_polkit_settings` too

File: src/lib/installation/security_settings.rb [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
```
Comment (dgdavid) [inline src/lib/installation/security_settings.rb:161]: Still thinking that this `require` could be moved to the top

File: test/lib/clients/security_finish_test.rb [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+  end
+
+  describe "#modes" do
+    it "runs on installation, update and autoinstallation" do
+      expect(subject.modes).to contain_exactly(:installation, :autoinst, :update)
+    end
+  end
+
+  # commented out to avoid build time dependency on firewall
+  xdescribe "#write" do
```
Comment (dgdavid) [inline test/lib/clients/security_finish_test.rb:29]: Is this a left over?

File: src/lib/installation/widgets/selinux_policy.rb [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+# current contact information at www.suse.com.
+# ------------------------------------------------------------------------------
+
+require "yast"
+require "cwm/common_widgets"
+
+module Installation
+  module Widgets
+    # Widget to set SELinux mode
+    class SelinuxPolicy < CWM::ComboBox
```
Comment (dgdavid) [inline src/lib/installation/widgets/selinux_policy.rb:26]: The widget actually manage the SELinux _mode_, not the _policy_. So, I'd rename it to be more accurate.

File: src/lib/installation/widgets/selinux_policy.rb [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+      def initialize(settings)
+        textdomain "installation"
+
+        @settings = settings
+      end
+
+      def label
+        # TRANSLATORS: SELinu Mode just SELinux is already content of frame.
```
Comment (dgdavid) [inline src/lib/installation/widgets/selinux_policy.rb:34]: NP: SELinu**x**

File: package/yast2-installation.changes [hunk 8]
```diff
@@ -1,3 +1,12 @@
+-------------------------------------------------------------------
+Wed Feb 24 20:56:30 UTC 2021 - Josef Reidinger <jreidinger@suse.com>
+
+- Add new security proposal and finish clients to replace firewall
+  one as the new clients contain also configuration for cpu
+  mitigation, policy kit default priviledges (jsc#SLE-15840)
+  and selinux policy (jsc#SLE-17307)
```
Comment (dgdavid) [inline package/yast2-installation.changes:7]: NP: selinux mode

**Review by dgdavid — CHANGES_REQUESTED** [review]
In general looks good to me. Just minor things plus two suggestions:

* to split a bit the SecurityFinish#write method and 
* to rename the SelinuxPolicy widget to SelinuxMode

However, the reason I'm putting the PR in "requests changes" mode is because there is a commented test unit and it is not clear to me if it is intentional or a left over. Could you please have a look?

Thanks!

File: package/yast2-installation.changes [hunk 9]
```diff
@@ -1,3 +1,12 @@
+-------------------------------------------------------------------
+Wed Feb 24 20:56:30 UTC 2021 - Josef Reidinger <jreidinger@suse.com>
+
+- Add new security proposal and finish clients to replace firewall
+  one as the new clients contain also configuration for cpu
+  mitigation, policy kit default priviledges (jsc#SLE-15840)
```
Comment (dgdavid) [inline package/yast2-installation.changes:6]: Typo: privileges, although in every place appears as _priviledges_

File: src/lib/installation/clients/security_finish.rb [hunk 10]
```diff
@@ -0,0 +1,181 @@
...
+
+        # Write down selinux configuration
+        settings.selinux_config.save
+
+        true
+      end
+
+    private
+
+      def write_polkit
```
Comment (dgdavid) [inline src/lib/installation/clients/security_finish.rb:90]: If I'm not wrong, this method is not being called.

File: test/lib/clients/security_finish_test.rb [hunk 11]
```diff
@@ -0,0 +1,171 @@
...
+    end
+  end
+
+  describe "#modes" do
+    it "runs on installation, update and autoinstallation" do
+      expect(subject.modes).to contain_exactly(:installation, :autoinst, :update)
+    end
+  end
+
+  describe "#write" do
```
Comment (dgdavid) [inline test/lib/clients/security_finish_test.rb:28]: Looks like we are not testing polkit related operations. In fact, I think that right now the new `write_polkit` private method is not being called.

**Review by dgdavid — CHANGES_REQUESTED** [review]

**Review by dgdavid — APPROVED** [review]
Looks good, thanks!

**Comment by dgdavid** [comment]
Tested manually in SLES-15-SP3

| Summary | Security client |
|-|-|
| Screenshot_sle15sp3-2_2021-02-25_16:30:27 | Screenshot_sle15sp3-2_2021-02-25_16:30:38 |

**Pull request merged**